docs/concepts/what-is-a-workflow-definition.md

```diff
@@ -11,13 +11,97 @@ A Workflow Definition is the code that defines the constraints of a Workflow Exe
 
 We strongly recommend that you write a Workflow Definition in a language that has a corresponding Temporal SDK.
 
+- [How to develop a Workflow Definition in Go](/docs/go/how-to-develop-a-workflow-definition-in-go)
+- [How to develop a Workflow Definition in Java](/docs/java/how-to-develop-a-workflow-definition-in-java)
+
+### Deterministic constraints
+
+A very important aspect of developing Workflow Definitions is making sure that the same Commands are emitted in the same order whenever a corresponding Workflow Function Execution (instance of the Function Definition) is re-executed.
+
+The execution semantics of a Workflow Execution include the re-execution of a Workflow Function.
+The use of Workflow APIs in the function generate the Commands.
+Commands tell the Cluster what Events to create and add to the Workflow Execution's Event History.
+Whenever a Workflow Function is executed, the Commands that are emitted are compared with the existing Event History.
+If a corresponding Event already exists within the Event History that maps to the generation of that Command in the same sequence, and some specific metadata of that Command matches with some specific metadata of the Event, then that Command is ignored.
+
+For example, using an SDKs "Execute Activity" API generates the [ScheduleActivityTask](/docs/concepts/what-is-a-command#scheduleactivitytask) Command.
+When this API is called upon re-execution, that Command is compared with the Event that is in the same location within the sequence.
+The Event in the sequence must be an [ActivityTaskScheduled](/docs/concepts/what-is-an-event/#activitytaskscheduled) Event, where the Activity Name and the Task Queue name are the same as what is in the Command.
+
+If a Command is generated that doesn't match what it needs to in the existing Event History, then the Workflow Execution will return a "non-deterministic" error.
+
+The following are the two reasons why a Command might be generated out of sequence or the wrong Command might be generated altogether:
+
+1. Code changes are made to a Workflow Definition that is in use by a running Workflow Execution.
+2. There is intrinsic non-deterministic logic (i.e. inline random branching logic).
+
+#### Non-deterministic code changes
```

Comment:
This header sounds like the code change itself is somehow done nondeterministically, where it's the code change that is causing nondeterminism.

Reply:
Hmmm... ya I see your point.
I can make it much more specific at the cost of length - 
I was thinking that the previous section would provide the clarifying context.